### PR TITLE
Task 1792621: [refactor] IOException is not used properly and severely polluted our code

### PR DESCRIPTION
**Pull request by wangmingliang-ms**
Almost all the fixed IOException are thrown from auth related code. 
User must signin first before they can use any feature of the plugin, so we can just throw a runtime exception to terminate user action if auth related problem found.

these are the original methods throwing IOException before fix (which finally pollutes >50 classes and hundreds of methods) 
image

### CODE REVIEW

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/Node.java [hunk 1]
```diff
@@ -328,12 +309,7 @@ public void onNodeDblClicked(Object context) {
     }
 
     public NodeAction getNodeActionByName(final String name) {
-        return Iterators.tryFind(nodeActions.iterator(), new Predicate<NodeAction>() {
-            @Override
-            public boolean apply(NodeAction nodeAction) {
-                return name.compareTo(nodeAction.getName()) == 0;
-            }
-        }).orNull();
+        return Iterators.tryFind(nodeActions.iterator(), nodeAction -> name.compareTo(nodeAction.getName()) == 0).orNull();
```
Comment (Flanker32) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/Node.java:312]: use `StringUtils.equals` ? As `compareTo` may more likely a sort method

File: Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/exception/AzureToolkitRuntimeException.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+package com.microsoft.azure.toolkit.lib.common.exception;
+
+import lombok.Getter;
+
+@Getter
+public class AzureToolkitRuntimeException extends RuntimeException {
+    private final String action;
+
+    public AzureToolkitRuntimeException(String error) {
```
Comment (Flanker32) [inline Utils/azuretools-core/src/com/microsoft/azure/toolkit/lib/common/exception/AzureToolkitRuntimeException.java:31]: error->message?

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java [hunk 3]
```diff
@@ -68,34 +64,36 @@ private AuthMethodManager(AuthMethodDetails authMethodDetails) {
         this.authMethodDetails = authMethodDetails;
         // initialize subscription manager when restore authentication
         if (this.authMethodDetails.getAuthMethod() != null) {
-            try {
-                getAzureManager().getSubscriptionManager().updateSubscriptionDetailsIfNull();
-            } catch (IOException e) {
-                // swallow exception
-            }
+            Objects.requireNonNull(getAzureManager()).getSubscriptionManager().updateSubscriptionDetailsIfNull();
```
Comment (Flanker32) [inline Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java:67]: I think we just need to check `getAzureManager() == null` and return, the constructor will be called in `LazyLoader` during plugin initialization, throw NPE here may failed the process.

**Review by Flanker32 — CHANGES_REQUESTED** [review]

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java [hunk 4]
```diff
@@ -68,34 +64,36 @@ private AuthMethodManager(AuthMethodDetails authMethodDetails) {
...
-    public AppPlatformManager getAzureSpringCloudClient(String sid) throws IOException {
-        if (getAzureManager() == null) {
-            throw new IOException(CANNOT_GET_AZURE_MANAGER);
+    public AppPlatformManager getAzureSpringCloudClient(String sid) {
+        final AzureManager manager = getAzureManager();
+        if (manager != null) {
+            return getAzureManager().getAzureSpringCloudClient(sid);
         }
-        return getAzureManager().getAzureSpringCloudClient(sid);
+        final String error = "Failed to initialize request";
```
Comment (Flanker32) [inline Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java:94]: duplicate with previous, create an exception `AzureToolkitNotSignedInException`?

File: Utils/azuretools-core/src/com/microsoft/azuretools/core/mvp/model/function/AzureFunctionMvpModel.java [hunk 5]
```diff
@@ -66,40 +53,42 @@ public static AzureFunctionMvpModel getInstance() {
...
 
-    public FunctionApp getFunctionById(String sid, String id) throws IOException {
-        FunctionApp app = getFunctionAppsClient(sid).getById(id);
-        if (app == null) {
-            throw new IOException(CANNOT_GET_FUNCTION_APP_WITH_ID + id); // TODO: specify the type of exception.
+    @NotNull
+    public FunctionApp getFunctionById(String sid, String id) throws AzureToolkitRuntimeException {
+        final FunctionApp app = getFunctionAppsClient(sid).getById(id);
+        if (Objects.isNull(app)) {
+            final String error = String.format("cannot find FunctionApp[%s] in subscription[%s]", id, sid);
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/core/mvp/model/function/AzureFunctionMvpModel.java:60]: Cannot
Confirm

**Review by andxu — DISMISSED** [review]

File: Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java (hunk 4 above)
Comment (wangmingliang-ms) [reply Utils/azuretools-core/src/com/microsoft/azuretools/authmanage/AuthMethodManager.java:94]: let's summarize such kind of common exceptions in the future.

**Pull request merged**